main/src/main/scala/sbt/internal/server/BuildServerProtocol.scala

```diff
@@ -315,6 +315,31 @@ object BuildServerProtocol {
     bspBuildTargetCompileItem := bspCompileTask.value,
     bspBuildTargetRun := bspRunTask.evaluated,
     bspBuildTargetScalacOptionsItem := scalacOptionsTask.value,
+    bspBuildTargetJVMRunEnvironment := Def.inputTaskDyn {
+      val s = state.value
+      val targets = spaceDelimited().parsed.map(uri => BuildTargetIdentifier(URI.create(uri)))
+      val workspace = bspFullWorkspace.value.filter(targets)
+      val filter = ScopeFilter.in(workspace.scopes.values.toList)
```

Comment:
I copied this from another request, how can I extract this to a helper function?

Reply:
`.value` and `.parsed` are macros and they can only be called inside `Def.setting`, `Def.task`, `Def.inputTask`.

So here you can try something like that:

```scala
bspBuildTargetJvmRunEnvironment := bspInputTask { (state, filter) =>
  Def.task {
    val items = bspBuildTargetJvmEnvironmentItem.result.all(filter).value
    val successfulItems = anyOrThrow(items)
    val result = JvmRunEnvironmentResult(successfulItems.toVector, None)
    state.respondEvent(result)
  }
}
```

where `bspInputTask` is:

```scala
private def bspInputTask[T](taskImpl: (State, ScopeFilter) => Def.Initialize[Task[T]]): Def.Initialize[InputTask[T]] = 
  Def.inputTaskDyn {
    val s = state.value
    val targets = spaceDelimited().parsed.map(uri => BuildTargetIdentifier(URI.create(uri)))
    val workspace = bspFullWorkspace.value.filter(targets)
    val filter = ScopeFilter.in(workspace.scopes.values.toList)
    taskImpl(s, filter)
  }
```

Reply:
Thanks! With `bspInputTask` I was able to refactor some of the tasks!